src/Database/Query.php

```diff
@@ -715,7 +715,7 @@ public function rightJoin($table, $conditions = [], $types = [])
      * The arguments of this method are identical to the `leftJoin()` shorthand, please refer
      * to that methods description for further details.
      *
-     * @param string|string[] $table The table to join with
+     * @param string|(string|\Cake\ORM\Query)[] $table The table to join with
```

Comment:
That syntax isn't supported in most IDEs, even the (not in PSR-5 anymore) `Type<...>` generic syntax has barely been adopted.

```suggestion
     * @param string|string[]|\Cake\Database\ExpressionInterface[] $table The table to join with
```

ps. don't forget `leftJoin()` and `_makeJoin()`.

Reply:
The proposed syntax can go into psalm-param tag on top if it provides additional info.

Reply:
Logically `string[]|\Cake\Database\ExpressionInterface[]` is wrong because the array may contain both strings and objects. It is not only array of strings or array of objects.

Reply:
@dereuromark, could you elaborate or give an example?

Reply:
@ndm2 https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#arrays

Reply:
@bancer Quoting back other resources that not make it any more valid.
We explained why you need to use `psalm-param` syntax on top for now for this syntax. You can find other pieces in the code here as reference already.

Reply:
@dereuromark, well @ndm2 stated that `(string|\Cake\ORM\Query)[]` syntax is not in PSR-5 anymore that is not true.

Reply:
I stated that the `Type<..>` syntax (`array<string|ExpressionInterface>`) is not in PSR-5 anymore, even though it had at least _some_ adoption.

Reply:
It doesn't matter if it is in or out. The facts remain: We cannot use it in tags yet, and thus we use the special tags on top.
The minor difference in meaning is therefore acceptable if that means we do not lose major IDE support here.
Please understand that this is why those special tags exist in the first place.

Reply:
The meaning is actually perfectly valid in this case, as `_makeJoin()` expects / makes use of exactly one array entry.